cpp/src/dictionary/dictionary_factories.cu

```diff
@@ -0,0 +1,186 @@
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/gather.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+
+namespace cudf
+{
+namespace detail
+{
+namespace
+{
+
+// Sort functor for column row positions.
+template <bool has_nulls = true>
+struct sort_functor
+{
+    column_device_view d_column;
+    __host__ __device__ bool operator()(size_type lhs_index, size_type rhs_index)
+    {
+        auto comparator = experimental::element_relational_comparator<has_nulls>{
+                          d_column, d_column, null_order::AFTER }; // put nulls at the end
+        auto result = experimental::type_dispatcher(d_column.type(), comparator,
+                                                    lhs_index, rhs_index);
+        return result == experimental::weak_ordering::LESS; // always sort ascending
+    }
+};
+
+// Copy functor used for locating and marking unique values.
+template <bool has_nulls = true>
+struct copy_unique_functor
+{
+    column_device_view d_column;
+    const int32_t* d_ordinals;
+    int32_t* d_indices;
+
+    __device__ bool operator()(int32_t idx)
+    {
+        if( idx==0 )
+        {   // first element is always unique
+            d_indices[0] = 0;
+            return true;
+        }
+        // check if adjacent elements match
+        auto lhs_index = d_ordinals[idx-1];
+        auto rhs_index = d_ordinals[idx];
+        auto result = !cudf::experimental::type_dispatcher( d_column.type(),
+                             experimental::element_equality_comparator<has_nulls>{d_column, d_column, true},
+                             lhs_index, rhs_index);
+        d_indices[idx] = static_cast<int32_t>(result); // convert bool to integer [0,1]
+        return result;
+    }
+};
+
+} // namespace
+} // detail
+
+/**
+ * @brief Create dictionary column from a column_view.
+ *
+ */
+std::unique_ptr<column> make_dictionary_column( column_view const& input_column,
+                                                rmm::mr::device_memory_resource* mr,
+                                                cudaStream_t stream)
+{
+    auto count = input_column.size();
+    auto execpol = rmm::exec_policy(stream);
+    auto d_view = column_device_view::create(input_column,stream);
+    auto d_column = *d_view;
+
+    // Example using a strings column:  [e,a,d,b,c,c,c,e,a]
+    //    row positions for reference:   0,1,2,3,4,5,6,7,8
+
+    rmm::device_vector<size_type> ordinals(count);
+    auto d_ordinals = ordinals.data().get();
+    thrust::sequence(execpol->on(stream), ordinals.begin(), ordinals.end()); // [0,1,2,3,4,5,6,7,8]
+    if( input_column.has_nulls() )
+        thrust::sort(execpol->on(stream), ordinals.begin(), ordinals.end(), detail::sort_functor<true>{d_column} );
+    else
+        thrust::sort(execpol->on(stream), ordinals.begin(), ordinals.end(), detail::sort_functor<false>{d_column} );
+    // output of sort:
+    //  ordinals: [1,8,3,4,5,6,2,0,7]  => these represent sorted strings as: [a,a,b,c,c,c,d,e,e]
```

Comment:
This can be ignored if using `drop_duplicates/lower_bound`.

Reply:
The sorted result here is only needed internally for this function. 
The`cudf::sorted_order` creates column which we really don't need.

Reply:
I understand it's internal, but Don't Repeat Yourself (DRY). 

You're correct that a `column` isn't necessary here, but this duplicates code and logic which introduces opportunities for bugs and inefficiencies that are already thought out and resolved in `sorted_order`. It's also more compact and descriptive because calling `sorted_order` tells the reader exactly what you're doing, instead of having to stare at the algorithm and figure out what's being done.